hathor/manager.py

```diff
@@ -334,6 +353,72 @@ def can_start_mining(self) -> bool:
             return True
         return self.connections.has_synced_peer()
 
+    def get_block_templates(self) -> List[BlockTemplate]:
+        """ Cached version of `make_block_templates`, cache is invalidated when latest_timestamp changes."""
+        cached_timestamp: Optional[int]
+        cached_block_template: List[BlockTemplate]
+        cached_timestamp, cached_block_template = getattr(self, '_block_templates_cache', (None, None))
+        if cached_timestamp == self.tx_storage.latest_timestamp:
+            return cached_block_template
+        block_templates = list(self.make_block_templates())
+        setattr(self, '_block_templates_cache', (self.tx_storage.latest_timestamp, block_templates))
+        return block_templates
+
+    def make_block_templates(self) -> Iterator[BlockTemplate]:
+        """ Makes block templates for all possible best tips as of the latest timestamp.
+
+        Each block template has all the necessary info to build a block to be mined without requiring further
+        information from the blockchain state. Which is ideal for use by external mining servers.
+        """
+        for parent_block_hash in self.tx_storage.get_best_block_tips():
+            yield self.make_block_template(parent_block_hash)
+
+    def make_block_template(self, parent_block_hash: bytes) -> BlockTemplate:
```

Comment:
I gotta say this whole method is very confusing to me. It's hard to understand the use of `generate_parent_txs` with `timestamp_abs_max` and later the calculations with weight decay.

Reply:
I'll try adding a clear explanation to the comments. But basically it boils down to narrowing the min and max timestamps, the absolute minimum would be the previous timestamp + 1, and absolute maximum limited by max time between blocks. The actual minimum is so that the block has a timestamp higher then its parent txs, and the actual maximum is when should the decay start (so we would want to update the weight).

Reply:
> The actual minimum is so that the block has a timestamp higher then its parent txs

I believe it still should be `previous timestamp + 1`. Worst case it just uses the same 2 parents as the last block.

Reply:
Fixed to include that +1 and also added more comments.